### DISCUSSION TITLE
Blocked from Flashing AI-Deck Due to CPX

### DISCUSSION THREAD

**Discussion by bhabas** [bitcraze/discussions#637] [Q&A] [answered]
It seems I made a mistake very similar to  (#542) where I inadvertantly  flashed a small test script to my AI-Deck, which repeatedly sends a message over CPX with no delay, causing problems (shown below but DO NOT RUN!). Since then, I cannot flash anything to the AI-Deck over Crazyradio and after flashing a new example script over JTAG, which works for a bit, my problem script will return.

```
#include "pmsis.h"
#include "stdio.h"
#include "cpx.h" 


void main_task(void)
{
    cpxPrintToConsole(LOG_TO_CRTP,"\n\n\t *** Camera Example ***\n\n");
    // pmsis_exit(-1); 
}

// PROGRAM ENTRY
int main(void)
{
    cpxInit();  
    return pmsis_kickoff((void *) main_task);
}
```

From my understanding, it seems like flashing over JTAG only uploads a script to the L2 RAM, and my problem script resides in the external flash memory and remains untouched ([Reference](https://www.bitcraze.io/documentation/repository/aideck-gap8-examples/master/infrastructure/flashing/) ). After reflashing the bootloader, it reconnects to the problem script, which then blocks communication over CPX so the Crazyradio route of flashing can't be completed.

Is there a way to clear out this memory and start with a clean-slate for the bootloader? 

Here's everything I have tried so far:

1. Flashing Example [Hello_World_Gap8](https://www.bitcraze.io/documentation/repository/aideck-gap8-examples/master/simple-examples/hello-world/) via Crazyradio:

   `cfloader flash examples/other/hello_world_gap8/BUILD/GAP8_V2/GCC_RISCV_FREERTOS/target.board.devices.flash.img deck-bcAI:gap8-fw -w radio://0/80/2M/E7E7E7E701`

   The Crazyflie tries to reboot into bootloader mode but jumps back into firmware mode, and the terminal hangs on "Reset to bootloader mode...".

2. Update Crazyflie and AI-Deck firmware via cfclient:

   I tried repeating the steps to  [update the CF and AI-Deck firmware](https://www.bitcraze.io/documentation/tutorials/getting-started-with-aideck/#preparing-the-ai-deck-and-crazyflie-firmware), but the update hangs on "Identifying deck to be updated" after updating the Crazyflie firmware and ESP32.

   After restarting cfclient, the console only outputs the problem script (shown below) with no access to Parameters, Battery Voltage, or data via Flight Control.

```
CPX: GAP8: 
	 *** Camera Example ***
CPX: GAP8: 
	 *** Camera Example ***
CPX: GAP8: 
	 *** Camera Example ***
CPX: GAP8: 
	 *** Camera Example ***
CPX: GAP8: 
```

3. Flashing the Gap8 Bootloader via Olimex ARM-USB-TINY-H JTAG:

   I can successfully reflash the Gap8 bootloader via JTAG, which returns control to cfclient. However, the same problem with hanging on "Identifying deck to be updated" comes back when I try to update the AI-Deck firmware again. If I reset the Crazyflie, the console output from the problem script returns as well.

4. Flash new script over JTAG:

   My next thought was to try flashing an example script over JTAG to overwrite the broken one. I tried doing this by going into the aideck-gap8-examples directory and running the command:
`
   docker run --rm -it -v $PWD:/module/ --device /dev/ttyUSB0 --privileged -P bitcraze/aideck tools/build/make-example examples/other/hello_world_gap8 flash`

   However, it seems this only overwrites the script in the L2 RAM and not the external flash memory where the problem script currently resides ([Reference](https://www.bitcraze.io/documentation/repository/aideck-gap8-examples/master/infrastructure/flashing/#system-architecture)).

   I can reboot the system, and it'll output the 'hello_world_gap8' script via the console. However, when I try to flash over Crazyradio:
`
   cfloader flash examples/other/hello_world_gap8/BUILD/GAP8_V2/GCC_RISCV_FREERTOS/target.board.devices.flash.img deck-bcAI:gap8-fw -w radio://0/80/2M/E7E7E7E701`

   The system reboots and hangs on 99% indefinitely:
```
   | 96% Writing to bcAI:gap8 deck memory
   / 97% Writing to bcAI:gap8 deck memory
   - 98% Writing to bcAI:gap8 deck memory
   \ 99% Writing to bcAI:gap8 deck memory
```

   When I disconnect and resume after completing any of the above steps, I return to the output from my problem script:

```
CPX: GAP8: 
	 *** Camera Example ***
CPX: GAP8: 
	 *** Camera Example ***
CPX: GAP8: 
	 *** Camera Example ***
CPX: GAP8: 
	 *** Camera Example ***
CPX: GAP8: 
```

**Comment by hmllr** [comment, marked as answer]
Hi :)
the flash command will write your gap8 program to flash, so will overwrite your "problem script". 
Also, the ESP pulls a pin to set the gap8 to the bootloader, so at least theoretically the code running on gap8 shouldn't matter (as long as it doesn't "kill" the esp somehow).
This means, your problem seems more related to the ESP than the gap8. Can you try updating the firmware on the ESP? You can use the cfclient (connect -> bootloader) with the latest release for this. Make sure that you see it updating the ESP (this will take some minutes), we have seen some issues there (which we are investigating but haven't solved yet) that the ESP is sometimes skipped - especially when a flowdeck is connected. 
So in short: detach all decks but the AI-deck, open the cfclien, go to connect->bootloader, program the newest release and see if this helps :)
Hanna

**Reply by bhabas** [reply]
Hi Hanna

It looks like the system has no problem updating the ESP using cfclient (connect -> bootloader). 
I attached a copy of the debug output from cfclient. I'm not sure if it's relevant but it seems to skip a gap8 update after being successful with updating the ESP.

`Skipping bcAI:gap8, no artifact for it in the .zip`

[CFClient_Debug_Output.txt](https://github.com/bitcraze/discussions/files/11100926/CFClient_Debug_Output.txt)

![AI_Deck_ESP_Update](https://user-images.githubusercontent.com/68243056/228535055-c5d72cfd-e19b-4719-92b1-af9203754a2d.png)

However, I still cannot upload over radio. The new script I recently uploaded over JTAG which prints 'Hello World' is still present (not the problem script) even after resetting (which is good) but when I try to upload another script over radio I get this output where it hangs on 99% indefinitely until I turn off the crazyflie:
```
\ 95% Writing to bcAI:gap8 deck memory
| 96% Writing to bcAI:gap8 deck memory
/ 97% Writing to bcAI:gap8 deck memory
- 98% Writing to bcAI:gap8 deck memory
\ 99% Writing to bcAI:gap8 deck memory
Got link error callback [Too many packets lost] in state [2]
Failed to flash: Failed to update deck bcAI:gap8
```


Edit:

I just reflashed the bootloader via JTAG and the problem script has returned with the output below filling up the cfclient console:

```
CPX: GAP8: 
CPX: GAP8: 
CPX: GAP8: 
CPX: GAP8: 
CPX: GAP8: 
CPX: GAP8: 
CPX: GAP8: 
	 *** Camera Example ***
CPX: GAP8: 
CPX: GAP8: 
CPX: GAP8: 
CPX: GAP8: 
CPX: GAP8: 
** Camera Example ***
mera Example ***
CPX: GAP8: 
CPX: GAP8: 
```

**Reply by hmllr** [reply]
This all looks correct to me. Skipping the gap8 update is normal, as there is no firmware for it in the release.

You reflashed the bootloader and different code than before is running now? Are you sure you reflashed the bootloader and not accidentally your script? I am sorry to ask so bluntly, but I am pretty sure the bootloader is jumping to a specific location which cannot be changed, so the code there should always be the same. EDIT: Actually, I am not sure if this is true, as the code is always run from RAM but stored in flash, so it might be copied over from the filesystem, making all kind of errors possible.

I am not sure what is happening - I saw the gap8 update hanging before, but never without any other deck and all up-to-date firmwares. 
I am also not sure how to erase the gap8. An a bit wild try might be to flash something huge, like the face detection example, over JTAG and afterwards the bootloader, so there is sure no valid firmware to jump to and see if this somehow helps to recover.

**Reply by bhabas** [reply]
No you're good, it's a valid question about the bootloader and things are definitely weird right now. I directly copy and pasted the commands from [here](https://www.bitcraze.io/documentation/tutorials/getting-started-with-aideck/#gap8-bootloader-for-the-aideck-1-0-or-aideck-1-1-rev-a-c-) and executed everything in a completely different directory so I don't see any possible connection with my previous script

It seems like the problem script is hidden in a small corner of the flash memory somewhere and keeps getting linked by the bootloader. If another script could write over that memory and corrupt the problem script that might help. I'll give that a shot now with the face detection example and see what comes up

**Reply by bhabas** [reply]
I was able to reflash with the FaceDetection example and tried a couple other ones for good measure but they all have the same behavior each time. 

1.  Flash example over JTAG:

```
docker run --rm -v ${PWD}:/module --device /dev/ttyUSB0 --privileged -P aideck-with-autotiler tools/build/make-example examples/image_processing/FaceDetection flash
docker run --rm -v ${PWD}:/module aideck-with-autotiler tools/build/make-example examples/image_processing/FaceDetection image
```
2. Cfclient console output seems normal even after resetting crazyflie multiple times (
[FaceDetect_Console_Output.txt](https://github.com/bitcraze/discussions/files/11101454/FaceDetect_Console_Output.txt))




3. Flash bootloader over JTAG
```
$ git clone https://github.com/bitcraze/aideck-gap8-bootloader.git
$ cd aideck-gap8-bootloader
$ docker run --rm -it -v $PWD:/module/ --device /dev/ttyUSB0 --privileged -P bitcraze/aideck /bin/bash -c 'export GAPY_OPENOCD_CABLE=interface/ftdi/olimex-arm-usb-tiny-h.cfg; source /gap_sdk/configs/ai_deck.sh; cd /module/;  make all image flash'
```

4. Reset crazyflie and console returns to the same problem output (
[Post_Bootloader_Output.txt](https://github.com/bitcraze/discussions/files/11101481/Post_Bootloader_Output.txt))

**Reply by hmllr** [reply]
Hmm. There is actually a file system. So maybe it has something to do with that, there @evoggy maybe knows more, this is not really my area of expertise - but I'll try anyway ;)
One idea would be to flash e.g. the classification example and set the filesystem offset to zero (https://github.com/bitcraze/aideck-gap8-examples/blob/master/examples/ai/classification/Makefile#L68) and then reflash the bootloader over it. Then I think we should really have broken the partition table... This is really strange though.

**Reply by bhabas** [reply]
That worked! I was just able to successfully upload the hello_world_gap8 example over radio and everything is now working!

For the purpose of documentation these were my steps:

1. In the classification demo [set the filesystem offset to zero](https://github.com/bitcraze/aideck-gap8-examples/blob/master/examples/ai/classification/Makefile#L68) and continue with [classification example directions](https://www.bitcraze.io/documentation/repository/aideck-gap8-examples/master/ai-examples/classification-demo/):
  Replace: `-DFS_PARTITIONTABLE_OFFSET=0x40000 with -DFS_PARTITIONTABLE_OFFSET=0x00000`

2. Place the `train` and `test` folders from `training_data_christmas_packet.zip` inside the training_data directory

3. Generate the CNN with only one epoch to reduce training time: 
    `python train_classifier.py --epochs 1`

4. Compile the classification script:
`docker run --rm -v ${PWD}:/module aideck-with-autotiler tools/build/make-example examples/ai/classification clean model build image`
5. Flash script to AI-deck via JTAG and validate console output:
```docker run --rm -v ${PWD}:/module --device /dev/ttyUSB0 --privileged -P aideck-with-autotiler tools/build/make-example examples/ai/classification flash```
[Classification_Output.txt](https://github.com/bitcraze/discussions/files/11102049/Classification_Output.txt)
6. Flash bootloader and validate console output:
    [Bootloader_Output.txt](https://github.com/bitcraze/discussions/files/11102054/Bootloader_Output.txt)
```
git clone https://github.com/bitcraze/aideck-gap8-bootloader.git
cd aideck-gap8-bootloader
docker run --rm -it -v $PWD:/module/ --device /dev/ttyUSB0 --privileged -P bitcraze/aideck /bin/bash -c 'export GAPY_OPENOCD_CABLE=interface/ftdi/olimex-arm-usb-tiny-h.cfg; source /gap_sdk/configs/ai_deck.sh; cd /module/;  make all image flash'
```

7. Compile and flash hello_world_gap8 example via cfloader
`python -m cfloader flash examples/other/hello_world_gap8/BUILD/GAP8_V2/GCC_RISCV_FREERTOS/target.board.devices.flash.img deck-bcAI:gap8-fw -w [CRAZYFLIE URI]`
[Hello_World_Gap8_Output.txt](https://github.com/bitcraze/discussions/files/11102081/Hello_World_Gap8_Output.txt)

**Reply by hmllr** [reply]
I am glad to hear it worked! And thanks for your detailed steps. I opened an issue for it here: https://github.com/bitcraze/aideck-gap8-examples/issues/126

**Reply by bhabas** [reply]
Thanks for your help!!

**Reply by softword-tt** [reply]
hi, thx for your detailed steps and i have a similiar issue with you (after a while loop cpxprint,  all cfloader steps stuck), but now i cannot get any JTAG, do you have any other ideas about how to clear the problem script without JTAG?

**Reply by knmcguire** [reply]
@softword-tt please start a new discussion and refer to this one. This discussion has already been resolved since an answer has been selected
